### ISSUE TITLE
Will the org front matter support some structure like an associative array?

### ISSUE THREAD

**Issue by xinii** [niklasfasching/go-org#39]
First of all, thanks very much for your development to support our emacs org-mode users to use Hugo more conveniently.
I'm comfortable writing Hugo posts via org now, but I'm a little bit confused about the handling of the associative array.

For example, if there is a YAML front matter like this,

```yaml
title: an amazing title
tags:
- tag1
- tag2
menu:
  example:
    parent: Example Topic
    weight: 1
```

About the keys of `title` and `tags`, if write them by org matter, they will be,
```org
#+title: an amazing title
#+tags[]: tag1 tag2
```

However, how to handle the following YAML matter which contains some structure like an associative array through org matter?

```yaml
menu:
  example:
    parent: Example Topic
    weight: 1
```

I tried some combinations that didn't seem to work. Is there any good way to solve it?

**Comment by niklasfasching** [comment]
As far as I know Org mode doesn't have support for lists and associative arrays - `#+TAGS[]` is already a hack, but it was simple enough (with limitations, e.g. there's no support for list items that contain whitespace). There's nothing as simple we can do for associative arrays that I'm aware of.

Why not just use yaml? As there's no use for associative arrays in Org mode itself and hugo allows you to use yaml (any other type of front matter really) front matter in Org mode files I don't think there's much value in pursuing support in org mode syntax.

**Comment by xinii** [comment]
> As far as I know Org mode doesn't have support for lists and associative arrays - `#+TAGS[]` is already a hack, but it was simple enough to [implement](https://github.com/gohugoio/hugo/blob/c66dc6c74fa3bbe308ccaade8c76071b49908129/parser/metadecoders/decoder.go#L229) (with limitations, e.g. there's no support for list items that contain whitespace). There's nothing as simple we can do for associative arrays that I'm aware of.
> 
> 
> 
> Why not just use yaml? As there's no use for associative arrays in Org mode itself and hugo allows you to use yaml (any other type of front matter really) front matter in Org mode files I don't think there's much value in adding Org mode native support.
> 
> 
> 
> TLDR: just use yaml
> 
> 
> 
> 

Thank you very much. Your answer solved my confusion well. I will use yaml as appropriate.

**Comment by pprevos** [comment]
Do you mean combining YAML into an org file? I worked around this by creating auxiliary files in the data folder.

**Comment by pprevos** [comment]
Hi Xinii,

You need to create a Toml/Taml or JSON file in the data folder, e.g. filename.yaml. You can then use a shortcode to extract the data, for example:

You shortcode: {{< shortcode-name "ref" >}}

In the shortcode, you can range over the data file:

{{ $ref := .Get 0 }} # Extract the shortcode parameter
{{ range .Site.Data.filename }} # Read the file
{{ if eq .ref $ref }} # Match the shortcode parameter
# your content
{{ end }}
{{ end }}

This is the manual: https://gohugo.io/templates/data-templates/

P:)

**Comment by xinii** [comment]
> Hi Xinii,
> 
> 
> 
> You need to create a Toml/Taml or JSON file in the data folder, e.g. filename.yaml. You can then use a shortcode to extract the data, for example:
> 
> 
> 
> You shortcode: {{< shortcode-name "ref" >}}
> 
> 
> 
> In the shortcode, you can range over the data file:
> 
> 
> 
> {{ $ref := .Get 0 }} # Extract the shortcode parameter
> 
> {{ range .Site.Data.filename }} # Read the file
> 
> {{ if eq .ref $ref }} # Match the shortcode parameter
> 
> # your content
> 
> {{ end }}
> 
> {{ end }}
> 
> 
> 
> This is the manual: https://gohugo.io/templates/data-templates/
> 
> 
> 
> P:)
> 
> 
> 
> 
> 
> 
> 
> 


I will try this, thank you very much! ^_^

**Comment by mdorman** [comment]
So I will observe that `#+PROPERTIES` is actually specifically described in the org manual as being key-value pairs: https://orgmode.org/manual/Property-Syntax.html#Property-Syntax

**Comment by niklasfasching** [comment]
Good point, i didn't know that. 

This gives us existing semantics for the first level of associative arrays - sadly the initial example shows that one level is not enough - e.g. `menu.example.parent = Example topic` would still not be representable like that afaict. I don't think org would be a usable frontmatter format even with this change.

If this is about supporting `#+PROPERTY` rather than associative arrays in general we should split the discussion into a new issue.
